<a href="https://colab.research.google.com/github/AdamStamm/bootcamp_alura/blob/main/Notebooks/Bootcamp_Data_Science_Modulo_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Apresentação

Esse notebook contempla os códigos aprendidos no **Módulo 3** do Bootcamp Data Science aplicada desenvolvido pela [Alura](http://alura.com.br). O foco de aprendizado é **Análise descritiva e testes estatíticos**.

#Importando bibliotecas.

In [1]:
import pandas as pd
from zipfile import ZipFile
 
#biblioteca de fluxo de entrada e saída
#importamos o pacote de leitura binária
from io import BytesIO
import requests

- A biblioteca IO foi feita para controle de fluxos de entrada/saída de dados, no módulo anterior usamos o pacote StringIO da io para capturar o fluxo de dados que estavam na memória (copy and paste). Agora usaremos para capturar o fluxo de dados binários.
Mais informações: https://docs.python.org/3/library/io.html
 
- A biblioteca requests permite que você faça requisições HTTP de maneira fácil sem a necessidade de uma API.
mais informações [aqui](https://requests.readthedocs.io/pt_BR/latest/user/quickstart.html) ou [aqui](https://www.digitalocean.com/community/tutorials/how-to-get-started-with-the-requests-library-in-python-pt)

#Aula 1 - Educação, saúde e tipos de dados.

##Carregando dados

Dessa vez carregaremos os dados de um arquivo zipado, sem descompacta-lo no computador, por maior comodidade upei o arquivo zip direto no meu [Github](https://github.com/AdamStamm/bootcamp_alura/).

In [2]:
#Fazendo requisição binária do arquivo zip do github
path = 'https://github.com/AdamStamm/bootcamp_alura/blob/main/dados/PeNSE_2015_AMOSTRA2.zip?raw=true'
r = requests.get(path)
zip_file = BytesIO(r.content)
 
# Leitura do arquivo zip
zip = ZipFile(zip_file) 
 
#pega o caminho da lista de arquivos
caminho_arquivo_zip = zip.namelist()[1]
 
dados = pd.read_csv(zip.open(caminho_arquivo_zip), sep=';')

In [3]:
dados.head()

ANOPESQ  PAIS  REGEOGR  VB00004  VB01001  ...  V0008  V0041  aluno  escola  turma
0     2015    76        1        1        2  ...      4      1      1       1    296
1     2015    76        1        1        2  ...      4      1      2       1    296
2     2015    76        1        1        1  ...      4      1      3       1    296
3     2015    76        1        1        1  ...      4      1      4       1    296
4     2015    76        1        1        1  ...      4      1      5       1    296

[5 rows x 181 columns]

Algumas propriedades do ZipFile

In [4]:
#mostrando o objeto que o ZipFile possui.  
zip

<zipfile.ZipFile file=<_io.BytesIO object at 0x7f153e549bf8> mode='r'>

In [5]:
#vendo a lista de arquivos contidos no .zip
zip.infolist()

[<ZipInfo filename='arquivos csv/' external_attr=0x10>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV' compress_type=deflate external_attr=0x20 file_size=7262170 compress_size=1235667>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=11420533 compress_size=1329882>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=5000642 compress_size=114324>,
 <ZipInfo filename='arquivos sas/' external_attr=0x10>,
 <ZipInfo filename='arquivos sas/pense_amostra2_aluno.sas7bdat' compress_type=deflate external_attr=0x20 file_size=24691712 compress_size=1957530>,
 <ZipInfo filename='arquivos sas/pense_amostra2_alunoescola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=45253632 compress_size=2461932>,
 <ZipInfo filename='arquivos sas/pense_amostra2_escola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=18105344 compress_size=332162>,
 <Zip

In [6]:
#outra forma de ver os arquivos, retornando uma lista
zip.namelist()

['arquivos csv/',
 'arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV',
 'arquivos sas/',
 'arquivos sas/pense_amostra2_aluno.sas7bdat',
 'arquivos sas/pense_amostra2_alunoescola.sas7bdat',
 'arquivos sas/pense_amostra2_escola.sas7bdat',
 'arquivos xlsx/',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNO.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNOESCOLA.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ESCOLA.xlsx',
 'Dicionario_PENSE_Microdados_AMOSTRA2.xls']

Propriedade do requests

In [7]:
#verificando a URL que a requests capturou
r.url

'https://raw.githubusercontent.com/AdamStamm/bootcamp_alura/main/dados/PeNSE_2015_AMOSTRA2.zip'

Iniciando análises iniciais

In [8]:
#verificando o tamanho do data frame
print(f'O conjunto tem {dados.shape[0]} linhas e {dados.shape[1]} colunas')

O conjunto tem 16556 linhas e 181 colunas


In [9]:
#visualizando colunas
dados.columns

Index(['ANOPESQ', 'PAIS', 'REGEOGR', 'VB00004', 'VB01001', 'VB01002',
       'VB01003', 'VB01004', 'VB01005', 'VB01006',
       ...
       'ESTRATO_EXP', 'ESTRATOGEOREG', 'PESO', 'V0006', 'V0007', 'V0008',
       'V0041', 'aluno', 'escola', 'turma'],
      dtype='object', length=181)

Percebemos que existem colunas com os nomes das variáveis difíceis de serem identificados, por essa razão é necessário estudarmos o dicionário de variáveis que acompanha o arquivo zip. Uma outra forma de abordagem é observar a forma desses dados como faremos a seguir.

In [10]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16556 entries, 0 to 16555
Columns: 181 entries, ANOPESQ to turma
dtypes: float64(7), int64(171), object(3)
memory usage: 22.9+ MB


Percebemos que existem 7 colunas do tipo float, 171 do tipo inteiro e 3 objects.

In [11]:
#visualizando dados por tipo
dados.select_dtypes(include='object')

VB17003 VB17004        PESO
0        37,1     157  299,735235
1        50,2     152  355,170081
2        48,2   161,4  299,735235
3        69,4   172,5  355,170081
4        78,9   167,5  299,735235
...       ...     ...         ...
16551    56,7   167,9  447,815539
16552    61,9   185,8   592,50744
16553    60,9   159,4  401,131512
16554    47,3   157,8  442,718091
16555    62,6   163,9   592,50744

[16556 rows x 3 columns]

Note que o tipo object em geral trata de strings, mas temos 3 colunas de tipo float sendo tratadas como strings, isso acontece por que no nosso idioma pontos flutuantes são debitados por vírgula, no formato americano são demorados por ponto, isso faz com que o código seja classificado como string na hora do carregamento dos dados. Para corrigir isso basta, na hora do carregamento dos dados, informar essa característica incluindo os parâmetros thousands para milhar e decimal para decimais.
 
Também é importante olhar o dicionário de dados, se verificarmos a coluna PESO refere-se ao peso *estatístico* dos dados e **não** do peso dos alunos.

- **VB17003:** Peso(Kg) do aluno
- **VB17004:** Altura do aluno
- **PESO:** Peso amostral para expansão.

In [12]:
#carregando conjunto de dados com os parâmetros adicionais (deixei uma nova
#célula apenas por questão didática)
dados = pd.read_csv(zip.open(caminho_arquivo_zip), sep=';',thousands='.', decimal=',')

In [13]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16556 entries, 0 to 16555
Columns: 181 entries, ANOPESQ to turma
dtypes: float64(10), int64(171)
memory usage: 22.9 MB


Note que já não temos mais as colunas do tipo objeto, vamos observar então as colunas do tipo float.

In [14]:
dados.select_dtypes(include='float64')

TEMPODESLOC  TEMPOTOTAL  TEMPOEDFIS  ...  VB17003  VB17004        PESO
0              0.0        15.0         0.0  ...     37.1    157.0  299.735235
1            120.0       510.0       260.0  ...     50.2    152.0  355.170081
2              0.0       585.0       130.0  ...     48.2    161.4  299.735235
3              0.0         0.0         0.0  ...     69.4    172.5  355.170081
4             30.0       165.0        90.0  ...     78.9    167.5  299.735235
...            ...         ...         ...  ...      ...      ...         ...
16551         30.0       400.0       110.0  ...     56.7    167.9  447.815539
16552         20.0       800.0       325.0  ...     61.9    185.8  592.507440
16553          0.0         0.0         0.0  ...     60.9    159.4  401.131512
16554         75.0       215.0        35.0  ...     47.3    157.8  442.718091
16555         50.0       485.0       110.0  ...     62.6    163.9  592.507440

[16556 rows x 10 columns]

*Dicionário:

- **VB17001**: Qual o peso do aluno?
- **VB17002**: Qual a altura do aluno?

OBS: Anteriormente tinhamos visto que peso e aluna do aluno eram a coluna VB17003 e VB17004, por que isso? As de final 01 e 02 são dados aferidos dos alunos e as 03 e 04 são dados imputados pelo sistema.

#Tipos de variáveis.

##Exemplos de tipos distintos

**Qualitativas**: São características que não possuem valores quantitativos, mas, ao contrário são definidas por diversas categorias, ou seja, representam uma classificação dos indivíduos. Podem ser de dois tipos.

- **Nominais**: São dados que não podem ser ordenados entre as categorias. Ex: cor dos olhos, doente/sadio, etc...

- **Ordinais**: São dados que possuem uma ordenação entre as categorias. Ex: estágio da doença (inicial, intermediário, terminal), mês de observação (janeiro, fevereiro...).

**Quantitativas**: São as características que podem ser medidas em uma escala quantitativa, ou seja, apresentam valores numéricos que fazem sentido. Também podem ser de dois tipos.

- **Contínuas**:  Características mensuráveis que assumem valores em uma escala contínua (na reta real do plano cartesiano), para quais valores fracionais fazem sentido. Ex: Peso (balança), altura (régua), tempo (relógio), pressão arterial, idade, etc...

- **Discretas**: São características mensuráveis que podem assumir apenas um número finito ou infinito contável de valores e, assim, somente fazem sentido valores inteiros. Geralmente são o resultado de contagens. Ex: Número de filhos, número de bactérias por litro de leite, número de cigarros fumados por dia.

##Desafio 1: Descompactar a pasta zipada utilizando outro método.

Para esse desafio irei fazer upload da pasta para o Google drive uma vez que o GitHub não permite a manipulação de dados sem que sejam dados commits. Portanto irei montar o drive e então trabalhar com o zip.
 
Após o drive montado irei utilizar comandos de Shell Script, ou seja, comandos bashrc. Os servidores colab utilizam por baixo dos panos o sistema Unix e isso permite esse tipo de utilização. Para enviarmos comandos para o sistema usamos o sinal de exclamação antes do comando.

In [15]:
#deixarei a celula comentada pois ainda irei trabalhar muito com esse notebook
#!unzip /content/drive/MyDrive/Dados_Bootcamp/PeNSE_2015_AMOSTRA2.zip -d /content/drive/MyDrive/Dados_Bootcamp

In [16]:
#visualizando arquivos extraídos
!ls /content/drive/MyDrive/Dados_Bootcamp

'arquivos csv'	  Dicionario_PENSE_Microdados_AMOSTRA2.xls
'arquivos sas'	  PeNSE_2015_AMOSTRA2.zip
'arquivos xlsx'


##Desafio 2: Estudar a base PeNSE.

In [17]:
#incluindo vídeo no notebook
from IPython.core.display import HTML
 
HTML('<center><iframe id="ytplayer" type="texto/html" width="500" height="300" \ src="https://www.youtube.com/embed/watch?v=z4BBCHZUvD4" frameborder="2"/></center>')

Os dados que serão trabalhados são da Pesquisa Nacional de Saúde do Escolar de 2015. Essa foi a terceira edição da PeNSE e foi a primeira que inclui alunos do ensino médio na faixa dos 13 aos 17 anos, as edições anteriores contemplavam apenas alunos até o nono ano (antiga oitava série).
 
Os estudantes são selecionados de um conjunto de escolas previamente definido (Amostra), e para cada um desses grupos escolares foi criado um plano tabular.
 
A Pesquisa é realizada 
graças à parceria do Ministério da Saúde com o Instituto Brasileiro de Geografia e Estatística e o apoio do Ministério da 
Educação. Ela faz parte da Vigilância dos Fatores de Risco e Proteção das Doenças Crônicas do Brasil. Em 2015, com a inclusão da amostra de estudantes de 13 a 17 anos, tornou a PeNSE 
comparável aos inquéritos internacionais.

##Desafio 3: Classificar algumas das perguntas do conjunto de dados como qualitativas e quantitativas.

**Quantitativas**
- Discretas:
 
>>- VB01010A (Contando com você, quantas pessoas moram na sua casa ou apartamento?)
- VB01019 (Quantos banheiros com chuveiro têm dentro da sua casa?)
- VB02001 (Nos últimos 7 dias, em quantos dias você comeu feijão?)
- VB02002 (Nos últimos 7 dias, em quantos dias você comeu salgados fritos?)
- VB02004A (Nos últimos 7 dias, em quantos dias você comeu pelo menos um tipo de vegetal?)
- VB03001A2 (Nos últimos 7 dias, em quantos você voltou a pé ou de bicicleta da escola?)

- Contínuas:

>>- VB17001: Qual o peso do aluno?
- VB17002: Qual a altura do aluno?
- TEMPODESLOC: Tempo de deslocamento diário acumulado pelo escolar, com deslocamento da casa para a escola e da escola para casa feito a pé ou de bicicleta.
- TEMPOEDFIS: Tempo médio acumulado em que o escolar fez atividades físicas.
- TEMPOEXTRA: Tempo médio diário acumulado do escolar com a prática de uma atividade física extraescolar como esportes, dança, ginástica, musculação...


**Qualitativas**
- Nominais:

>>- VB01023 (Você estuda em regime integral?)
- VB01024 (Você estuda em regime de internato?)
- VB01026 (Quando terminar o ciclo/curso que você está frequentando atualmente, você pretende?)
- VB01006 (Você mora com sua mãe?)
- VB01007 (Você mora com seu pai?)
- VB01013 (Na sua casa tem telefone fixo?)
- VB01014 (Você tem celular?)
- VB01015A (Na sua casa tem computador?)
- VB01016 (Você tem aceso à internet em sua casa?)
- VB01018 (Alguém que mora na sua casa tem carro?)

- Ordinais:

>>- VB01004 (Mês de aniversário?)
- VB01005 (Em que ano nasceu?)
- VB01021 (Em que série/ano você está?)
- VB01022 (Em que turno você estuda?)
- VB01025 (Qual o grau de escolaridade mais elevado que pretendo concluir?)
- VB01008A (Que nível de ensino (grau) sua mãe estudou ou estuda?
- VB03002A1 (Quando você vai para a escola a pé ou de bicicleta, quanto tempo você gasta?
- VB06006 (Quantos amigos seus usam drogas?)

#Aula 2 - Dados qualitativos e tabela de frequência.

Agora vamos iniciar uma análise da frequência das repostas do questionário.

In [18]:
#Verificando a frequência de respostas por sexo do aluno.
dados['VB01001'].value_counts()

1    8287
2    8269
Name: VB01001, dtype: int64

In [19]:
#verificando a porcentagem
dados['VB01001'].value_counts(normalize=True)*100

1    50.054361
2    49.945639
Name: VB01001, dtype: float64

A relação a alunos do sexo masculino e feminino que responderam ao questionário é bem equilibrada. Isso acontece por que essa pergunta faz parte da estratégia de amostragem da pesquisa. No entanto, existem outras perguntas que são independentes dessa estratégia. Como por exemplo a relação de satisfação do aluno com relação a seu corpo (variável VB11007) conforme analisaremos abaixo.

In [20]:
frequencia = dados['VB11007'].value_counts()

A resposta 2 é a meais respondida, seguido da 1 e da 4... Mas note que é difícil da gente visualizar isso, portanto vamos ao dicionário.

In [21]:
percentual = dados['VB11007'].value_counts(normalize=True)*100
percentual

2     41.797536
1     27.832810
4     14.049287
3     10.914472
5      4.312636
99     1.093259
Name: VB11007, dtype: float64

In [22]:
#Criando a lista de respostas conforme dicionário
respostas = {1: 'Muito Satisfeito',
             2: 'Satisfeito',
             3: 'Indiferente',
             4: 'Insatisfeito',
             5: 'Muito insatisfeito',
             99: 'Não informado'}
#Criando um data frame para melhor análise dos resultados acima.
dist_freq = pd.DataFrame({'Frêquencia': frequencia, 'Percentual': percentual})
dist_freq.rename(index=respostas,inplace=True)
dist_freq.rename_axis('Sentimento em relação ao seu corpo', axis=1, inplace=True)
#arredondando os valores percentuais
dist_freq['Percentual'] = dist_freq['Percentual'].round(2)
dist_freq

Sentimento em relação ao seu corpo  Frêquencia  Percentual
Satisfeito                                6920       41.80
Muito Satisfeito                          4608       27.83
Insatisfeito                              2326       14.05
Indiferente                               1807       10.91
Muito insatisfeito                         714        4.31
Não informado                              181        1.09

Percebemos que a maioria dos alunos se sentem satisfeitos em relação ao próprio corpo.

##Desafio 4: Criar uma função para gerar as tabelas de freqûencia e, além disso, fazer a análise de outras informações.

In [23]:
def tabela_frequencia(dados, dicionario=None, titulo_frequencias=None):
    '''
    Retorna uma tabela de frequencia baseada na coluna e no dataframe recebido.

    Parametros:
        dados (series): Series contendo os dados selecionados
        dicionario (dict): Dicionário das respostas/index
        titulo_frequencias (str): Título para a coluna index.
    Retornos:
        dados (DataFrame): Tabela de frequencias das respostas
    '''

    frequencia = dados.value_counts()
    percentual = dados.value_counts(normalize=True) * 100
    percentual = percentual.round(2)
    dados = pd.DataFrame({"Frequencia": frequencia, "Percentual": percentual})
    if dicionario:
      dados.rename(index=dicionario, inplace=True)
    if titulo_frequencias:
      dados.rename_axis(titulo_frequencias,axis=1,inplace=True)

    return dados

In [24]:
img_corp = {1: 'Muito Importante',
            2: 'Importante',
            3: 'Indiferente',
            4: 'Insatisfeito(a)',
            5: 'Muito insatisfeito(a)',
            99: 'Não informado'}
test_function = tabela_frequencia(dados['VB11007'],img_corp, 'Grau') 

In [25]:
test_function

Grau                   Frequencia  Percentual
Importante                   6920       41.80
Muito Importante             4608       27.83
Insatisfeito(a)              2326       14.05
Indiferente                  1807       10.91
Muito insatisfeito(a)         714        4.31
Não informado                 181        1.09

#Aula 3 - Frequência em dados contínuos e Pandas CrossTab

Agora vamos criar esse tipo de tabela de frequência mas com dados contínuos. Trabalharemos com o VB17004 a altura dos estudantes.

In [26]:
dados['VB17004']

0        157.0
1        152.0
2        161.4
3        172.5
4        167.5
         ...  
16551    167.9
16552    185.8
16553    159.4
16554    157.8
16555    163.9
Name: VB17004, Length: 16556, dtype: float64

Veja que não temos classes pré-definidas nesse caso, como construir, portanto, uma tabela de frequência para esse tipo de informação?
Uma das formas é criarmos variáveis conforme segue.

In [27]:
#Criando a variável IMC (VB17003: peso, VB17004: altura), 
#o IMC é o resultado da divisão do peso pela a altura ao quadrado
#como a altura está em centimetros e o IMC faz a conta por metros iremos dividir
#por 100 a altura.
dados['IMC'] = dados['VB17003'] / ((dados['VB17004']/100)**2)
dados['IMC']

0        15.051321
1        21.727839
2        18.502907
3        23.322831
4        28.122076
           ...    
16551    20.113223
16552    17.930782
16553    23.968489
16554    18.995343
16555    23.303246
Name: IMC, Length: 16556, dtype: float64

Criando as classes

In [28]:
classes = [0, 18.5, 25, 30, 35, 60]
labels = ['abaixo', 'adequado', 'sobrepeso', 'obesidade grau I', 'obesidade grau II']

dados['CLASSE_IMC'] = pd.cut(x = dados['IMC'],
       bins = classes,
       labels = labels,
       include_lowest = True)

dados[['IMC','CLASSE_IMC']]

IMC CLASSE_IMC
0      15.051321     abaixo
1      21.727839   adequado
2      18.502907   adequado
3      23.322831   adequado
4      28.122076  sobrepeso
...          ...        ...
16551  20.113223   adequado
16552  17.930782     abaixo
16553  23.968489   adequado
16554  18.995343   adequado
16555  23.303246   adequado

[16556 rows x 2 columns]

Entendendo o intervalo ao remover a opção labels.

In [29]:
classes = [0, 18.5, 25, 30, 35, 60]
labels = ['abaixo', 'normal', 'sobrepeso', 'obesidade grau I', 'obesidade grau II']

dados['CLASSE_IMC'] = pd.cut(x = dados['IMC'],
       bins = classes,
       labels = labels,
       include_lowest = True)

dados[['IMC','CLASSE_IMC']]

IMC CLASSE_IMC
0      15.051321     abaixo
1      21.727839     normal
2      18.502907     normal
3      23.322831     normal
4      28.122076  sobrepeso
...          ...        ...
16551  20.113223     normal
16552  17.930782     abaixo
16553  23.968489     normal
16554  18.995343     normal
16555  23.303246     normal

[16556 rows x 2 columns]

Note que o intervalo é do mais baixo (lowest) como aberto `(` e o superior como intervalo fechado `]`

In [30]:
dist_freq = tabela_frequencia(dados['CLASSE_IMC'])
dist_freq

Frequencia  Percentual
normal                   9188       55.50
abaixo                   4975       30.05
sobrepeso                1760       10.63
obesidade grau I          488        2.95
obesidade grau II         145        0.88

Note que 30% dos estudantes estão com o peso abaixo do normal, interesse fato a ser investigado.

##Regra de Sturges

Até agora vimos como definir classificações para frequências, no caso do IMC podemos verificar essas classificações por dados na literatura e outras fontes. Mas e quando não temos esses dados? Quantas classes devemos colocar? Ou seja, qual o número adequado de classes para um conjunto de dados? É o que veremos agora.